Fix SM300D2 sensor component routines so they correctly read the sensor output

jamesbraid opened the pull request:
# What does this implement/fix? 

Fixes the SM300D2 sensor to work more reliably, the previous code seems to fail because it reads 17 bytes from the serial port expecting them to be aligned to the buffer where the sensor appears to just be streaming data

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <https://github.com/esphome/issues/issues/2273>


## Test Environment

- [x] ESP32
- [ ] ESP8266

## Example entry for `config.yaml`:


```yaml
esphome:
  name: ${device_name}
  platform: ESP32
  board: nodemcu-32s

# logger:
#   baud_rate: 0

uart:
  rx_pin: GPIO17
  baud_rate: 9600

sensor:
  - platform: uptime
    name: Uptime

  - platform: wifi_signal
    name: Wifi Signal
    update_interval: 60s

  - platform: sm300d2
    co2:
      name: "CO2"
    formaldehyde:
      name: "Formaldehyde"
    tvoc:
      name: "TVOC"
    pm_2_5:
      name: "PM2.5"
    pm_10_0:
      name: "PM10"
    temperature:
      name: "Temperature"
    humidity:
      name: "Humidity"
    update_interval: 10s


```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

jamesbraid committed:
sm300d2: return humidity and temperature to 1 decimal place

oxan commented:
Hi, thanks for your contribution! Could you please elaborate a bit more on why these changes are necessary? Your code changes behaviour in quite a few ways:
* It no longer flushes before reading data (which should be a no-op, as nothing writes to the UART bus).
* It retries reading the message 3 times if the checksum doesn't match.
* Doesn't check whether the second byte of the message is `0x02`.
* Skips over any extra bytes read from the UART bus before the message header.

From your commit message, it seems the last thing is the change you actually wanted. That can also be achieved by just inserting this at the beginning of the `update()` method (which is imo cleaner):

```cpp
uint8_t peeked;
while (this->available() > 0 && this->peek_byte(&peeked) && peeked != 0x3C)
  this->read();
```

Maybe the other things are also necessary, but if so, it'd be nice to know why :)

jamesbraid commented:
Ah, nice! I'll try that approach and see if it works, appreciate the response!

jamesbraid committed:
sm300d2: use consistent log format for PM10 output

Log in the same format as other sensor measurements.

jamesbraid committed:
sm300d2: check for header byte before reading message

Make sure have a valid sensor message before reading the message from
the serial port. Remove useless flush() calls - sensor does not have a
TX pin connected.

Fixes: 8e93735861c785f5883a908022d4418cc413f1de

jamesbraid commented:
> From your commit message, it seems the last thing is the change you actually wanted. That can also be achieved by just inserting this at the beginning of the update() method (which is imo cleaner):

Thanks for the feedback, that is indeed a lot cleaner and works well in initial testing. I'll leave it running for stress testing overnight.

For the other comments.

- the flush() calls were removed because as you say, there's no writes happening, and the sensor doesn't even have a TX serial pin connected from the ESP. 

- the previous version returned CRC errors ~2% of the time so I wanted to retry on CRC error instead of returning a missing data point for that read cycle. May have been a function of my previous implementation rather than actual errors. Will monitor and see if that is still required.

poptix commented:
These patches resolve the remainder of the problems I've had with the sm300d2 sensor. It's practically unusable without them.

oxan reviewed (approved):
@jamesbraid I assume from your lack of further comments that my suggestions didn't cause any issues? :)

If so, this can be merged.

OttoWinter reviewed (approved):
Thanks!
